Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -901,6 +899,19 @@ extension TaskHandler: ChannelDuplexHandler {
         }
     }
 
+    private func writeBodyPart(context: ChannelHandlerContext, part: IOData, promise: EventLoopPromise<Void>) {
+        switch self.state {
+        case .idle:
+            self.actualBodyLength += part.readableBytes
+            context.writeAndFlush(self.wrapOutboundOut(.body(part)), promise: promise)
+        default:
+            let error = HTTPClientError.writeAfterRequestSent
+            promise.fail(error)
```

Comment:
This can't be quite right (and I think we should have a test for this). We're calling out to the user here _before_ we change our state. So we may get user code "sandwiched in" here.

Reply:
I've re-ordered things, not sure yet how to test it though, I'll have to get access to a handlers internal state... Could you elaborate on "sandwiched" code a bit?

Reply:
@artemredkin sure. If you do `promise.fail(...)`, then the promise's `whenFailure` and other code will run _inline_ here. So we detected an error (ie. we _should_ be in state `.endOrError`) but we call out _before_ we entered the `.endOrError` state. So if the user for example uses another method that does a state check, we potentially have an issue because their code runs before the `self.state = .endOrError` so they won't see it. That could lead us to make wrong decisions.

Reply:
ah, that is so obvious now :) I'll think about how to test this

Reply:
@Lukasa could you pick up from Johannes here? I've added a test to verify that if delegate is called here, state is changed. I haven't figured out how to attach anything to that promise though... It seems that I need to have access to `channel.write(HTTPClient.request, promise)`, and I don't think anything in the code can attach to it